### PR TITLE
F6 PR1 — D-6 PB-2 client auth + GrantContext.authenticatedClient + public-client PKCE mandatory (v0.5.1)

### PR DESCRIPTION
**Pull request by y1o1**
## Summary

F6 PR1 of v0.5.1 hotfix series — **BREAKING**. Closes RFC 6749 §2.3 client-authentication holes (PB-2) + closes Area-2-NEW-1 (discovery `"none"` was a future lie) + partial SF-7 (PKCE/S256 mandatory for public clients per RFC 9700).

Spec: `auth/.claude/superpowers/specs/2026-05-05-d6-pb2-client-schema-redesign.md`

### Core changes

- **`Client.tokenEndpointAuthMethod`** required discriminator (`"client_secret_basic" | "client_secret_post" | "none"`); `clientSecret` optional, `.superRefine` enforces "secret iff confidential, forbidden iff `none`"; all `Client` fields `readonly`.
- **`AuthenticatedClient` + `GrantContext.authenticatedClient`** — populated by `clientAuthMw`, consumed by grants (no more body-spoofable `body.client_id`).
- **`clientAuthMw` full refactor** — public-client path (Codex M0 `allowPublicClients` opt-in, `/token` only), per-method enforcement (Codex M1), Basic+body conflict detection (Codex M4), issuer-driven WWW-Authenticate realm.
- **`/oauth/token`** wires `clientAuthMw` ahead of grant dispatch; rate limit moved before client auth to bound DoS amplification through repository lookups.
- **`/oauth/authorize`** PKCE/S256 mandatory for `tokenEndpointAuthMethod === "none"` (RFC 9700 §2.1.1) — operator `pkce.required` config no longer consulted for public clients.
- **`refreshToken.mts`** ctx.authenticatedClient gate + replace ALL downstream raw-body sinks (Codex M3 — finalAudience, grantPolicy clientId, new AT+RT aud+azp, authorizedParty); binding gate `(claims.azp ?? aud) === authenticatedClient.clientId`; new tokens emit `azp` explicitly.
- **`authorization.mts`** removed in-grant client_secret block (RFC 6749 §2.3 belongs at the route); replaced ≥7 raw-body `client_id` identity sinks (Codex M2 — token aud/azp, ID-token aud/azp, grant-policy, RP registration, logout-metadata findById); canonical binding `codeData.client_id === authenticatedClient.clientId`.
- **Token Exchange grant** trusts `ctx.authenticatedClient` when present (Basic-auth callers don't repeat credentials in body); standalone wiring path retains body-credential gate. Public clients refused regardless of route.

### Codex review rounds

3 rounds of `codex review` between commits — final round Approved.

| # | Finding | Fix |
|---|---|---|
| R1 P1 | clientAuthMw shared with `/introspect` admitted public clients via `client_id` only (RFC 7662 §2.1 violation) | `allowPublicClients` opt-in option, `/token` sets true, `/introspect` default false |
| R1 P2 | Basic-auth Token Exchange broke (in-grant body-secret check) | Grant trusts `ctx.authenticatedClient` when present |
| R2 P1 | Round-1 Token Exchange fix still required body.client_id (earlier required-field guard rejects null) | Resolve `clientId = body.client_id ?? ctx.authenticatedClient?.clientId` |
| R3 P2 | Round-2 fix accepted malformed body.client_id (string[]) via silent null fallback | Reject with 400 invalid_request when present-but-not-a-string |

### Fixture migration

- 50+ `GrantContext` literals updated across `oauth/`, `oauth-token-exchange/` test files (added `authenticatedClient` field).
- `InMemoryClientRepository` test fixtures + create-app + standalone `clients.yaml.example` updated with `tokenEndpointAuthMethod`.
- Smoke test registers a confidential test client and sends Basic auth on `/oauth/token` calls.

### Test count

`a1d5dfee` baseline 2054 → HEAD 0cdc4ae4 ~2089 (+35 RED tests):

- Group A schema discriminator (7) — A-1..A-6
- Group B clientAuthMw paths (15+) — B-1..B-9, B-method-1..3, B-conflict, repository fail-closed, edge cases, issuer realm, allowPublicClients gate (3)
- D-6 binding-gate tests in `refreshToken.test.mts` + `authorization.test.mts`
- R-legacy-azp legacy aud-only RT path
- Token Exchange D-6 ctx.authenticatedClient flow (4) — Basic-auth happy path, body.client_id omitted, malformed array, public-client rejection

All 2089 tests green; biome clean (4 pre-existing warnings unchanged); typecheck clean across all packages.

### Migration

`config/clients.yaml` requires `tokenEndpointAuthMethod` per client. See CHANGELOG.md.

## Test plan

- [x] `pnpm -r run build` — all packages compile
- [x] `pnpm -r run test` — 2089 tests pass
- [x] `pnpm exec biome check packages/...` — no new warnings
- [x] Codex `review --commit HEAD` Approved
- [ ] Copilot review on PR (pending)
- [ ] Squash-merge to develop after Copilot ✅

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/oauth/src/middleware/clientAuth.mts [hunk 1]
```diff
@@ -50,109 +75,270 @@ function formUrlDecode(s: string): string {
...
+	// Backward-compat: F1 D-4 callers passed the Logger directly. Accept either
+	// form to avoid forcing every call site to migrate at once. New call sites
+	// SHOULD pass the options object so they can supply `issuer` for the
+	// realm parameter.
+	const opts: ClientAuthMiddlewareOptions =
+		typeof loggerOrOptions === "object" && "warn" in loggerOrOptions
+			? { logger: loggerOrOptions as Logger }
+			: (loggerOrOptions as ClientAuthMiddlewareOptions);
+	const logger: Logger = opts.logger ?? consoleLogger;
+	const realm = opts.issuer && opts.issuer.length > 0 ? opts.issuer : "oauth";
```
[the remark this answers is not in the document]
Comment (y1o1) [reply packages/oauth/src/middleware/clientAuth.mts]: Fixed in d7896dfe — `issuer` is validated against an RFC 3986 + sub-delim-safe character set before being embedded into the realm quoted-string; any unsafe input (`"`, `\`, CR, LF, control bytes) falls back to the literal `oauth` realm. New RED test in clientAuth.test.mts asserts that an issuer carrying CRLF + Set-Cookie injection produces `Basic realm="oauth"`.

File: packages/oauth/src/middleware/clientAuth.mts [hunk 2]
```diff
@@ -50,109 +75,270 @@ function formUrlDecode(s: string): string {
...
+ *    body sent to a `client_secret_basic` client) reject with `invalid_client`
+ *    400/401, matching RFC 6749 §5.2.
+ * 2. If both Basic and body credentials are present and the `client_id` (or
+ *    `client_secret`) values disagree, reject with `invalid_client` 401 —
+ *    prevents credential-confusion attacks where an attacker pins a victim's
+ *    Basic header and supplies their own body credentials, or vice versa.
+ * 3. `WWW-Authenticate: Basic realm="<issuer>"` is emitted only when the
+ *    failed attempt was Basic (per RFC 7235 §2.1 the header advertises a
+ *    challenge, and a `client_secret_post` client should not be told to retry
+ *    over Basic). Public-client failures emit no challenge for the same reason.
```
[the remark this answers is not in the document]
Comment (y1o1) [reply packages/oauth/src/middleware/clientAuth.mts]: Fixed in d7896dfe — JSDoc rewritten to spell out both cases: WWW-Authenticate is emitted (a) when the failed attempt was Basic and (b) when no credentials were supplied at all (so the caller knows Basic is an accepted retry transport). Behavior unchanged; the doc now matches.

File: packages/oauth/src/routes.mts [hunk 3]
```diff
@@ -176,87 +194,108 @@ export const createOAuthRouter = async (
...
+						userAgent: req.get("user-agent"),
+						details: { grant_type },
+					});
+					return res.status(result.status).json(result.tokens);
+				}
+				const errorBody: Record<string, unknown> = { error: result.error };
+				if (result.errorDescription) errorBody.error_description = result.errorDescription;
+				if (result.status === 401) {
+					res.set("WWW-Authenticate", "Bearer");
+				}
```
[the remark this answers is not in the document]
Comment (y1o1) [reply packages/oauth/src/routes.mts]: Fixed in d7896dfe — removed the unconditional `WWW-Authenticate: Bearer` injection on /oauth/token 401 responses. RFC 6750 §3 applies to protected resource servers, not the token endpoint, and after D-6 the new `ctx.authenticatedClient === null` branch returns 401, so the Bearer header was clobbering clientAuthMw's upstream Basic challenge. Existing routes.test.mts test updated to assert no WWW-Authenticate is set on handler-returned 401.

File: packages/oauth-token-exchange/src/grant.mts [hunk 4]
```diff
@@ -101,19 +127,68 @@ export function createTokenExchangeGrant(deps: TokenExchangeDependencies): Grant
...
+				if (clientSecret === null) {
+					return {
+						result: {
+							status: 401,
+							error: "invalid_client",
+							errorDescription: "client_secret is required",
+						},
+					};
+				}
+				client = await clientRepository.authenticate(clientId, clientSecret);
```
[the remark this answers is not in the document]
Comment (y1o1) [reply packages/oauth-token-exchange/src/grant.mts]: Fixed in d7896dfe — wrapped the standalone-wiring `clientRepository.authenticate(clientId, clientSecret)` in try/catch and convert repository failures to `503 temporarily_unavailable`, mirroring the authenticated-client branch. New RED test in grant.test.mts asserts that a throwing repository surfaces 503 instead of an unhandled 500.

File: templates/standalone/src/__tests__/smoke.test.mts [hunk 5]
```diff
@@ -89,11 +89,43 @@ const config: AppConfig = {
...
 const testRepositoriesModule = defineModule({
 	name: "test:repositories",
 	provides: {
-		clientRepository: () => new InMemoryClientRepository(new Map()),
+		// `InMemoryClientRepository` constructor takes `Map<string, ClientEntry>`
+		// where `ClientEntry = z.infer<typeof ClientEntrySchema>` (output type) —
+		// fields with `.default(...)` are required in the output even though the
+		// input shape only requires the discriminator + secret. Tests supply the
+		// input shape and let the schema fill defaults at parse time, hence the
+		// cast at the construction boundary.
```
[the remark this answers is not in the document]
Comment (y1o1) [reply templates/standalone/src/__tests__/smoke.test.mts]: Fixed in d7896dfe — comment rewritten. The literal supplies every defaulted field explicitly (no cast, no schema-time fill), and the comment now reflects the actual code path.
